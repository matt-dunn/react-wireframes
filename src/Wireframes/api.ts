/** !
 * Copyright (c) 2019, Matt Dunn
 *
 * @author Matt Dunn
 */

import { ComponentType, ReactNode } from "react";

import { getWireframeAnnotation, updateWireframeAnnotation } from "./utils";

type APIOptions = {
  updater: (annotations: WireframeAnnotations) => void;
  highlightNote?: (annotation?: WireframeAnnotation) => void;
}

export type WireframeAnnotationOptions = {
  title: ReactNode;
  description: ReactNode;
}

export type WireframeAnnotation = {
  id: number;
  Component: ComponentType<any>;
  count: number;
  options: WireframeAnnotationOptions;
}

export type WireframeAnnotations = WireframeAnnotation[];

type OpenCallback = (isOpen: boolean) => void;

export type WireframeAnnotationAPI = {
  setOptions: (options: APIOptions) => APIOptions;
  getAnnotations: () => WireframeAnnotations;
  register: (Component: ComponentType<any>, options: WireframeAnnotationOptions) => WireframeAnnotation;
  unregister: (Component: ComponentType<any>) => void;
  highlightNote: (Component: ComponentType<any> | undefined) => void;
  setOpen: (isOpen: boolean) => boolean;
  onOpen: (cb: OpenCallback) => {unregister: () => void};
  isOpen: () => boolean;
  setParentReference: (id?: number) => void;
  getParentReference: () => number | undefined;
}

export type withWireframeAnnotationProps = {
  annotationId?: number;
}

export function API(defaultOptions?: APIOptions): WireframeAnnotationAPI {
  let annotations: WireframeAnnotations = [];
  let apiOptions: APIOptions = defaultOptions || {} as APIOptions;
  let parentReference: number | undefined;

  let openCallbacks: OpenCallback[] = [];
  let isOpen = false;

  return {
    setOptions: (options) => {
      apiOptions = options;
      return apiOptions;
    },
    getAnnotations: () => annotations,
    register: (Component, options) => {
      const annotation = getWireframeAnnotation(annotations, Component);

      if (annotation) {
        const updatedAnnotation = {
          ...annotation,
          count: annotation.count + 1,
        };

        annotations = updateWireframeAnnotation(annotations, annotation, updatedAnnotation);

        /* istanbul ignore else */
        if (apiOptions.updater) {
          apiOptions.updater(annotations);
        }

        return updatedAnnotation;
      }

      const newAnnotation = {
        id: annotations.length + 1,
        Component,
        count: 1,
        options,
      };

      annotations = [...annotations, newAnnotation];

      if (apiOptions.updater) {
        apiOptions.updater(annotations);
      }

      return newAnnotation;
    },
    unregister: (Component) => {
      const annotation = getWireframeAnnotation(annotations, Component);

      if (annotation) {
        if (annotation.count > 1) {
          annotations = updateWireframeAnnotation(annotations, annotation, {
            ...annotation,
            count: annotation.count - 1,
          });
        } else {
          annotations = annotations.filter(c => c !== annotation);
        }

        if (apiOptions.updater) {
          apiOptions.updater(annotations);
        }
      }
    },
    highlightNote: Component => apiOptions && apiOptions.highlightNote && apiOptions.highlightNote(getWireframeAnnotation(annotations, Component)),
    setOpen: (open) => {
      if (open !== isOpen) {
        isOpen = open;

        openCallbacks.forEach(cb => cb(isOpen));
      }

      return isOpen;
    },
    onOpen: (cb) => {
      openCallbacks = [...openCallbacks, cb];

      return {
        unregister: () => {
          openCallbacks = openCallbacks.filter(c => c !== cb);
        },
      };
    },
    isOpen: () => isOpen,
    setParentReference: (id) => {
      parentReference = id;
    },
    getParentReference: () => parentReference,
  };
}
