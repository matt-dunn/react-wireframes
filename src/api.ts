/** !
 * Copyright (c) 2019, Matt Dunn
 *
 * @author Matt Dunn
 */

import { ComponentType, ReactNode } from "react";
import { remove } from "lodash";

type APIOptions = {
    updater: (components: WireFrameComponents) => void;
    highlightNote?: (component?: WireFrameComponent) => void;
}

export type WireFrameComponentOptions = {
    title: ReactNode;
    description: ReactNode;
}

export type WireFrameComponent = {
    id: number;
    Component: ComponentType<any>;
    count: number;
    options: WireFrameComponentOptions;
}

export type WireFrameComponents = WireFrameComponent[];

type OpenCallback = (isOpen: boolean) => void;

export type WireFrameAnnotationAPI = {
    setOptions: (options: APIOptions) => APIOptions;
    getComponents: () => WireFrameComponents;
    register: (Component: ComponentType<any>, options: WireFrameComponentOptions) => WireFrameComponent;
    unregister: (Component: ComponentType<any>) => void;
    highlightNote: (Component: ComponentType<any> | undefined) => void;
    setOpen: (isOpen: boolean) => boolean;
    onOpen: (cb: OpenCallback) => {unregister: () => void};
    isOpen: () => boolean;
}

export const API = function API(defaultOptions?: APIOptions): WireFrameAnnotationAPI {
  let components: WireFrameComponents = [];
  let apiOptions: APIOptions = defaultOptions || {} as APIOptions;

  let openCallbacks: OpenCallback[] = [];
  let isOpen = false;

  const getComponent = (Component: ComponentType<any> | undefined): WireFrameComponent | undefined => Component && components.find(c => c.Component === Component);

  return {
    setOptions: (options) => {
      apiOptions = options;
      return apiOptions;
    },
    getComponents: () => components,
    register: (Component, options) => {
      const component = getComponent(Component);

      if (component) {
        component.count += 1;

        if (apiOptions.updater) {
          apiOptions.updater(components);
        }

        return component;
      }

      const newComponent = {
        id: components.length + 1,
        Component,
        count: 1,
        options,
      };

      components = [...components, newComponent];

      if (apiOptions.updater) {
        apiOptions.updater(components);
      }

      return newComponent;
    },
    unregister: (Component) => {
      const component = getComponent(Component);

      if (component) {
        component.count -= 1;

        if (component.count === 0) {
          components = remove(components, c => c.Component !== Component);
        }
      }

      if (apiOptions.updater) {
        apiOptions.updater(components);
      }
    },
    highlightNote: Component => apiOptions && apiOptions.highlightNote && apiOptions.highlightNote(getComponent(Component)),
    setOpen: (open) => {
      isOpen = open;
      openCallbacks.forEach(cb => cb(isOpen));
      return isOpen;
    },
    onOpen: (cb) => {
      openCallbacks = [...openCallbacks, cb];
      return {
        unregister: () => {
          openCallbacks = remove(openCallbacks, c => c !== cb);
        },
      };
    },
    isOpen: () => isOpen,
  };
};
