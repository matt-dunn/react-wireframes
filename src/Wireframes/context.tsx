/** !
 * Copyright (c) 2019, Matt Dunn
 *
 * @author Matt Dunn
 */

import React, { createContext, ReactNode } from "react";

import { WireframeAnnotationAPI } from "./api";

export const WireframeAnnotationContext = createContext<WireframeAnnotationAPI | undefined>(undefined);

type WireframeProviderProps = {
  children: ReactNode;
  api: WireframeAnnotationAPI;
}

export const WireframeProvider = ({ children, api }: WireframeProviderProps) => (
  <WireframeAnnotationContext.Provider value={api}>
    {children}
  </WireframeAnnotationContext.Provider>
);
