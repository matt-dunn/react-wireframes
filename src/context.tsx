/** !
 * Copyright (c) 2019, Matt Dunn
 *
 * @author Matt Dunn
 */

import React, { createContext, ReactNode } from "react";

import { WireFrameAnnotationAPI } from "./api";

export const WireFrameAnnotationContext = createContext<WireFrameAnnotationAPI | undefined>(undefined);

type WireFrameProviderProps = {
  children: ReactNode;
  api: WireFrameAnnotationAPI;
}

export const WireFrameProvider = ({ children, api }: WireFrameProviderProps) => (
  <WireFrameAnnotationContext.Provider value={api}>
    {children}
  </WireFrameAnnotationContext.Provider>
);
