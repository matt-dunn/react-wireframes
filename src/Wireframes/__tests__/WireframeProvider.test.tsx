/** !
 * Copyright (c) 2019, Matt Dunn
 *
 * @author Matt Dunn
 */

import React from "react";
import { mount } from "enzyme";

import { WireframeAnnotationAPI } from "../api";

import { WireframeProvider } from "../context";

describe("Wireframe: WireframeProvider", () => {
  it("should render children", () => {
    const api = {} as WireframeAnnotationAPI;

    const wrapper = mount(
      <WireframeProvider api={api}>
        <div>Child</div>
      </WireframeProvider>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("should render children", () => {
    const api = {
      setParentReference: jest.fn(),
    } as unknown as WireframeAnnotationAPI;

    const apiInner = {
      setParentReference: jest.fn(),
    } as unknown as WireframeAnnotationAPI;

    const wrapper = mount(
      <WireframeProvider api={api}>
        <WireframeProvider api={apiInner} annotationId={12}>
          <div>Child</div>
        </WireframeProvider>
      </WireframeProvider>,
    );

    expect(wrapper).toMatchSnapshot();

    expect(api.setParentReference).not.toHaveBeenCalled();

    expect(apiInner.setParentReference).toBeCalledWith({
      id: 12,
      api,
    });
  });
});
