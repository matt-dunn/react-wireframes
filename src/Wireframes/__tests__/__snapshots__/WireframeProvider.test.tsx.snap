// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wireframe: WireframeProvider should render children 1`] = `
<WireframeProvider
  api={Object {}}
>
  <div>
    Child
  </div>
</WireframeProvider>
`;
